### ISSUE TITLE
Support for `torch.autograd.functional.jacobian` in TorchScript

### ISSUE THREAD

**Issue by Linux-cpp-lisp** [pytorch/pytorch#69741]
### 🚀 The feature, motivation and pitch

As I've mentioned in [previous issues ](https://github.com/pytorch/pytorch/issues/68621#issuecomment-978140881) the forward pass of our scientific models includes using the autograd engine to compute the gradient of a predicted quantity. This is happily well supported in TorchScript.

We now want to be able to deploy a model whose forward pass involves the _Jacobian_ of a predicted quality in the forward pass. This does not appear to be supported:
```
import torch

@torch.jit.script
def f(x):
    return (x + x[..., ::-1]).square()

@torch.jit.script
def g(x):
   return torch.autograd.functional.jacobian(
       func=f,
       inputs=(x,)
   )

g(torch.randn(4))
```
gives various TorchScript language errors.

### Alternatives

One option is to explicitly, in TorchScript, call `torch.autograd.grad` _N_ times (one for each entry in the output tensor) and then reassemble the Jacobian using `torch.cat`. This would only involve calls already supported in TorchScript and some simple loops, but has the worse-case performance of calling the autograd engine _N_ independent times.

Much preferable would be some kind of interface to PyTorch's own implementation that would allow taking advantage of `vmap` and `vectorize=True` for the Jacobian.

**Comment by albanD** [comment]
This is not planned from the autograd side. You can see details on the torchscript bigger picture here: https://dev-discuss.pytorch.org/t/next-steps-for-pytorch-compilers/309

**Comment by Chillee** [comment]
Depending on what you need this for, you could feasibly trace through it with `make_fx` from functorch and then Torchscript the result.

**Comment by Linux-cpp-lisp** [comment]
@albanD just to make sure I understand, you aren't planning to _remove_ support for these sorts of features from TorchScript, are you? The Python inference approach is cool but we really need the low overhead of TorchScript for smaller models and fast inference...

**Comment by Linux-cpp-lisp** [comment]
@Chillee : that's an interesting thought that I've considered when `functorch` first came out, but I haven't looked back at the current state since, which I should. Would `make_fx` of a vectorized Jacobian give an FX graph containing only a single set of instructions that are vectorized?

For me, the biggest issue with that is that our input shapes are _almost_, but never fully, static. It's a graph neural network, so all the feature dimensions **except** for the number of nodes and number of edges are constant.  Padding could be a solution, but feels quite wasteful... there's also no way for us to determine reliably a maximum number of edges, let alone a tight bound.

One solution I can imagine is to trace with `make_fx` using an input where the number of nodes/edges is some moderately large prime number, do if a few times, and then replace in the FX graph all occurrences of those static shapes with a dynamic shape derived from the input. Given that the only changing dimensions for us are batch dimensions, this may have some chance of working.

**Comment by Chillee** [comment]
@Linux-cpp-lisp Yeah, Torchscript will remain around for the foreseeable future (and we're continuing to invest into it as a performance backend). The primary implication of that post is that Torchscript isn't going to substantially expand its "frontend capture" capabilities. But... it's still possible to say, Torchscript the result of a `make_fx` graph, and LazyTensor also often goes through Torchscript when compiling.

Regarding the dynamic-shapes issues... it depends on your function. We're still planning on working on tracing dynamic shapes, but that won't be ready for some time.

As for your solution, that's one possibility... If you have an example I could potentially take a look at it. 

For what it's worth, here is an example of using our compilation features (successfully!) with jacobian computation: https://github.com/pytorch/functorch/issues/328#issuecomment-990286263

In the short term, I unfortunately don't see any way forward with compiling something like this *except* through an approach like the one we're using in functorch or an analogous one with the same shape-specializing characteristics.

**Comment by Linux-cpp-lisp** [comment]
@Chillee glad to hear— if I understand what you're saying, TorchScript is basically going to become more of a "target" for these various ways of capturing the behavior of a model, rather than substantially expanding its capacity to directly `torch.jit.script` compile models involving more advanced Python/PyTorch features?

> Regarding the dynamic-shapes issues... it depends on your function. We're still planning on working on tracing dynamic shapes, but that won't be ready for some time.

Cool! Looking forward to seeing more about this in the future.

> As for your solution, that's one possibility... If you have an example I could potentially take a look at it.
> ...
> In the short term, I unfortunately don't see any way forward with compiling something like this except through an approach like the one we're using in functorch or an analogous one with the same shape-specializing characteristics.

I suspect that — in my usecase — the dynamic part of the shapes is so simple (always dim 0, always the batch that passes through other operations unchanged) that something like this ugly hack I propose will work... I'll let you know if I have a chance to play around with this and share what I do.

-------

I don't mean to hijack my own thread, but while I have you here on this topic: @Chillee do you expect that I would see performance gains by using `grad` and `make_fx` to "unfold" the backwards pass in advance of TorchScript compilation? (Compared to a TorchScript model that includes a `torch.autograd.grad` call.)

The one thing I can think of is that this would let me use inference mode, since there would be no autograd in normal inference, but I'm not sure what a difference that makes. Does autograd itself (buillding, traversing, and destroying graphs) add significant overhead?

Thanks!

**Comment by Chillee** [comment]
> TorchScript is basically going to become more of a "target" for these various ways of capturing the behavior of a model, rather than substantially expanding its capacity to directly torch.jit.script compile models involving more advanced Python/PyTorch features?

Yes, essentially (that's my understanding).

> do you expect that I would see performance gains by using grad and make_fx to "unfold" the backwards pass in advance of TorchScript compilation?

Yeah, totally possible, although depends a lot on your actual operations/what hardware you're running on. In other words, how overhead bound is your problem? 

In the linked issue, there was a 3x difference in performance (by tracing it out), but that's 1. a small model, 2. running on CPU (where overheads don't really get hidden).

**Comment by Linux-cpp-lisp** [comment]
Wow 3x is a lot, even if considering that it's small and CPU (which is a performance regime we are also curious about, so it's interesting even without trying to extrapolate to CUDA.)

What is the difference between using `aot_function` and `torch.jit.script`ing `make_fx` of `grad`?

**Comment by Chillee** [comment]
Basically, `aot_function` consists of 3 steps:

1. Tracing the forward and jvp of a function jointly using `make_fx`.
2. Partitioning the joint graph into the forwards and the backwards graph.
3. Doing whatever you want with the forwards and backwards graphs (such as passing it torchscript).
4. We wrap up the new forwards and backwards passes up in an `autograd.Function` so that it can be used just like the original function.

Essentially, the underlying tracing mechanism (`make_fx`) is shared, and then you can do similar things (in terms of lowering the FX graphs to Torchscript and such). But, the extra steps of 2 and 4 allow us to integrate `aot_function` into regular eager-mode, even in the presence of autograd.

For example, imagine you had something like

```
y = foo(x)
z = bar(y)
z.sum().backward()
```

In this case, assume that `bar` isn't compilable (perhaps it has control flow, or whatever). In this case, you're not really able to use `make_fx(grad(foo))` to compile the forwards and backwards of `foo`. But you can use `aot_function` to do so.

There's some other more minor things as well, such as a compilation cache for `aot_function` and some utility functions to do stuff with modules in `aot_module`, but the above is the primary methodological difference.

I'll say that ideally, we (functorch-wise) are intending for `aot_function/module` to be for eager-mode purposes, and I suppose something like `make_fx` is more for the "trace out your whole program use cases" (which it sounds like you're in).

**Comment by Rullec** [comment]
> Aha I see, yes in that case `make_fx` sounds like it's much more appropriate for my purposes. Thanks for the info!

Sorry for bothering. I'm quite a newbie in developing torch. ;-) At this moment, I *encounter the same problem* as you when I try to record the jacobian computation. Would you please give me some insights? Thanks in advance!

# 1. My situation
I conduct `torch.jit.trace/script` over `javrec / jacobian` function from `functorch`, but failed.

Google search leads me here. I watched all your discussion carefully, but would still like to ask a question: How can I use `torch.jit.script/trace` to record a `jacobian` computation, no matter by `make_fx` or other method? Or would you be happy to show us an exmaple?

(Btw, It seems I cannot find the documentation of `make_fx`, it's quite annoying.)

# 2. Code
FYI, My minimal reproduce case here:
```python
import torch 
from functorch import jacrev, vmap
import torch.nn as nn

class Arch(nn.Module):
    def __init__(self, input_dim, output_dim) -> None:
        super().__init__()

        self.fc1 = nn.Linear(input_dim, 8)
        self.fc2 = nn.Linear(8, output_dim)
        self.tanh1 = nn.Tanh()
        
    def forward(self, x):
        x = self.fc1(x)
        x = self.tanh1(x)
        x = self.fc2(x)
        return x

dim = 2
batch = 4

net = Arch(dim, dim)
x = torch.randn(batch, dim)
primal = x.clone().requires_grad_()

# 1. succ: calculate the jacobian of the net by `jacrev` from functorch
jac_mat = vmap(jacrev(net.forward))(x)

# 2. failed: fail to trace it
module = torch.jit.trace(vmap(jacrev(net.forward)), x)

```

**Comment by zou3519** [comment]
Try the following:
```py
import torch 
from functorch import jacrev, vmap, make_fx
import torch.nn as nn

class Arch(nn.Module):
    def __init__(self, input_dim, output_dim) -> None:
        super().__init__()

        self.fc1 = nn.Linear(input_dim, 8)
        self.fc2 = nn.Linear(8, output_dim)
        self.tanh1 = nn.Tanh()
        
    def forward(self, x):
        x = self.fc1(x)
        x = self.tanh1(x)
        x = self.fc2(x)
        return x

dim = 2
batch = 4

net = Arch(dim, dim)
x = torch.randn(batch, dim)
primal = x.clone().requires_grad_()

# 1. succ: calculate the jacobian of the net by `jacrev` from functorch
jac_mat = vmap(jacrev(net))(x)

# 2. trace via make_fx
graph = make_fx(vmap(jacrev(net)))(x)
print(graph.code)
```

**Comment by Rullec** [comment]
> Try the following:
> 
> ```python
> import torch 
> from functorch import jacrev, vmap, make_fx
> import torch.nn as nn
> 
> class Arch(nn.Module):
>     def __init__(self, input_dim, output_dim) -> None:
>         super().__init__()
> 
>         self.fc1 = nn.Linear(input_dim, 8)
>         self.fc2 = nn.Linear(8, output_dim)
>         self.tanh1 = nn.Tanh()
>         
>     def forward(self, x):
>         x = self.fc1(x)
>         x = self.tanh1(x)
>         x = self.fc2(x)
>         return x
> 
> dim = 2
> batch = 4
> 
> net = Arch(dim, dim)
> x = torch.randn(batch, dim)
> primal = x.clone().requires_grad_()
> 
> # 1. succ: calculate the jacobian of the net by `jacrev` from functorch
> jac_mat = vmap(jacrev(net))(x)
> 
> # 2. trace via make_fx
> graph = make_fx(vmap(jacrev(net)))(x)
> print(graph.code)
> ```

Thanks for your reply! Actually it works. I can indeed get some code by `make_fx`.But its output is a function which has a `self` as an parameter. 

I'm a newbie on `torch`, and as the lackness of doc of `make_fx`, I really have no idea about how to convert it to a `torchscript` object (just same as the output of `torch.jit.trace`). 

Would you be happy to give me some insight on such a convertion?

**Comment by nawazishkhan1-nk** [comment]
@Rullec  @Chillee Any updates on this? I am stuck in a similar problem.

**Comment by janeyx99** [comment]
TorchScript has been deprecated in favor of torch.export as of 2.11. https://pytorch.org/blog/pytorch-2-11-release-blog/ Closing this issue as we are not further developing on it--please ping if you believe this issue was erroneously closed.
